### ISSUE TITLE
Create API for better performance

### ISSUE THREAD

**Issue by ericcornelissen** [ericcornelissen/shescape#856]
Relates to #273

## Summary

With the current [library API](https://github.com/ericcornelissen/shescape/blob/a9ab694cf03568bfeb8662f78d7720d3eb9130df/docs/api.md) there is no way around evaluating several conditions plus `which(1)` the shell in order to escape/quote the given argument. If invoking a command is at least somewhat common - e.g. in response to a HTTP request - this can become quite, and more importantly unnecessarily, expensive.

Through a new API it would be possible to provide an interface that is similarly expressive but supports evaluating `if`-statements and a `which(1)` invocation only once. E.g. using a [`class`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes) or (functional) builder pattern.

Conditions to consider:

- Selecting the escape/quote function for the given shell.
- Getting the platform helpers (relatedly, loading modules unrelated to the current OS unnecessarily).

## Timeline _(or rather, version-line)_

~~Experimentally, this new API may appear in the [v1](https://github.com/ericcornelissen/shescape/tree/d8dcbecf9bd3101559dad80fe2eda2869b8e89d9) release line. However, it will probably be unstable as I don't expect to get the API right in one go.~~ This will most likely not happen due to the increased urgency of releasing v2 due to #960.

Starting with [v2](https://github.com/ericcornelissen/shescape/milestone/1) this would probably become the "default" API. Depending on what this new API will look like, the current purely functional API may be kept available (adjusting for breaking changes in the release) in v2 as well.

I suspect, given that - again - the API may be unstable initially, that this change will be the reason to have `-rc` release for v2. But we'll see how/if an experimental release in v1 works.

**Comment by ericcornelissen** [comment]
## Proposal A

An API that is nearly identical to the existing API, just with a class (which must be instantiated) providing the escape/quote functionality. This class would evaluate the necessary conditions and whatnot in the constructor so that calling any of the escape/quote functions is more performant.

```ts
type ShellOption = boolean | string | undefined;

interface EscapeOptions {
  readonly interpolation?: boolean;
  readonly shell?: ShellOption;
}

interface QuoteOptions {
  readonly shell?: ShellOption;
}

class Shescape {
  escape(arg: string, options?: EscapeOptions): string;
  escapeAll(args: string[], options?: EscapeOptions): string[];
  quote(arg: string, options?: QuoteOptions): string;
  quoteAll(arg: strings[], options?: QuoteOptions): string[];
}
```

The main benefit of this approach is the similarity to the [v1 API](https://github.com/ericcornelissen/shescape/blob/2e41727cb6a4143c08836eb10f0f8338d392bc68/docs/api.md). Migration could essentially be as easy as:

```diff
   import { execFile } from "node:child_process";
-  import * as shescape from "shescape";
+  import { Shescape } from "shescape";
+  const shescape = new Shescape();

   // ...

   function safeEcho(args, cb) {
     const safeArgs = shescape.escapeAll(args);
     execFile("echo", safeArgs, options, cb);
   }
```

**Comment by ericcornelissen** [comment]
This was first available in [2.0.0-rc1](https://www.npmjs.com/package/shescape/v/2.0.0-rc1).
